go/vt/vtgate/evalengine/func.go

```diff
@@ -643,3 +645,48 @@ func (ta *typeAggregation) result() sqltypes.Type {
 	}
 	return sqltypes.VarChar
 }
+
+type builtinCeil struct{}
+
+func (builtinCeil) call(env *ExpressionEnv, args []EvalResult, result *EvalResult) {
+	inarg := &args[0]
+	argtype := inarg.typeof()
+	if inarg.isNull() {
+		result.setNull()
+		return
+	}
+
+	if sqltypes.IsIntegral(argtype) {
+		inarg.makeFloat()
+		result.setInt64(int64(math.Ceil(inarg.float64())))
+	} else if sqltypes.Decimal == argtype {
+		num := inarg.decimal()
+		if num.Cmp(decimal.NewFromInt(math.MaxInt32)) == 1 || num.Cmp(decimal.NewFromInt(math.MinInt32)) == -1 {
+			if num.Sign() == 1 {
+				result.setDecimal(num.Add(decimal.NewFromInt(1)), 0)
+			} else {
+				result.setDecimal(num.Add(decimal.NewFromInt(-1)), 0)
+			}
+		} else {
+			floatpart, _ := num.Float64()
+			result.setInt64(int64(math.Ceil(floatpart)))
+		}
```

Comment:
Can you explain a little bit what are you trying to do here? It seems like you're using different rounding modes for decimals that are larger than INT32_MAX or smaller than INT32_MIN -- otherwise you're just rounding the decimal as if it were a float...

I don't think this is exactly MySQL's behavior. Are you attempting an optimization here? I.e. doing the float rounding if it would always be safe to do so? I'm afraid this won't behave like you're expecting it to... Converting a decimal to a `float64` is super expensive! The comparisons against `math.MaxInt32` and `math.MaxInt32` are also expensive, particularly because they allocate two temporary decimals, and the `num.Float64` would tell you anyway if the conversion to `float64` was safe to perform.

I think this is a case where the simplest code would would also be the safest: you can add (or subtract) one from the decimal and truncate the decimals, and that should give you a good ceil... but here's an even better tip: if you check the implementation of our `decimal` package, you'll see that we've adapted an existing arbitrary-precision decimal implementation to behave like MySQL's. Maybe the original implementation has a `Ceil` function you could port over? :)

Reply:
I am confused with the mysql behaviors here. For example, when I test the value like `1.5`
``` bash
/home/hwj/cncf/vitess/go/vt/vtgate/evalengine/integration/comparison_test.go:710: 
different results: DECIMAL(2); mysql response: INT64(2) (local collation: binary; mysql collation: binary)
      query: SELECT CEIL(1.5) (SIMPLIFY=true)
```
Then I checked the type of the `1.5`, it is not the `FLOAT`, but `Demical`. Do I misunderstand the process here?

Reply:
From the MySQL docs:

> For exact-value numeric arguments, the return value has an exact-value numeric type. For string or floating-point arguments, the return value has a floating-point type.

It seems like rounding a DECIMAL, which is an exact value numeric type, will return INT64 if the result is small enough to fit, otherwise a DECIMAL.

Reply:
It it tricky to detect when to return `INT`, because:

``` go 

"9223372036854775810.4", // bigger than MAXINT64, return DECIMAL
"7223372036854775010.1", // bigger than MAXINT32, return DECIMAL
"2147483648.1", // bigger than MAXINT32, return INT

```

Reply:
All you have to check is whether the result of the rounding fits in `int64` or not -- there are no intermediate types. If it fits, it is returned as `INT64`, if not as `DECIMAL`.

Reply:
When the type of args is `DECIMAL`, but the result is `INT64`. How could we return the true type in the `type` function. My current solution:
``` go
func (builtinCeil) call(env *ExpressionEnv, args []EvalResult, result *EvalResult) {
	inarg := &args[0]
	argtype := inarg.typeof()
	if inarg.isNull() {
		result.setNull()
		return
	}

	if sqltypes.IsIntegral(argtype) {
		result.setInt64(inarg.int64())
	} else if sqltypes.Decimal == argtype {
		num := inarg.decimal()
		num = num.Ceil()
		intnum, isfit := num.Int64()
		if isfit {
			inarg.makeSignedIntegral()
			result.setInt64(intnum)	
		} else {
			result.setDecimal(num, 0)
		}
	} else {
		inarg.makeFloat()
		result.setFloat(math.Ceil(inarg.float64()))
	}
}

func (builtinCeil) typeof(env *ExpressionEnv, args []Expr) (sqltypes.Type, flag) {
	if len(args) != 1 {
		throwArgError("CEIL")
	}
	t, f := args[0].typeof(env)
	if sqltypes.IsIntegral(t) {
		return sqltypes.Int64, f
	} else if sqltypes.Decimal == t {
		return sqltypes.Decimal, f
	} else {
		return sqltypes.Float64, f
	}
}
```

How could we catch the new return type, like `1.5`?

Reply:
Huh, that's a good question. MySQL's behavior here is really bad... Our `typeof` helpers cannot (should not, really) return variable types for the same input, so we're gonna have to hack it and do something different to MySQL... Let's assume for now that `typeof` returns `Decimal` for `Decimal`. That'll give some bad `typeof` results from the tests, and we'll have to ignore them, but I think that's the best we can do for now.

Reply:
The `inarg.makeSignedIntegral()` line is superfluous by the way. You can remove it.